libmamba/include/mamba/solver/solution.hpp

```diff
@@ -218,35 +203,72 @@ namespace mamba::solver
                 action
             );
         }
+
+        template <typename Range>
+        auto packages_to_omit_impl(Range& actions)
+        {
+            namespace views = std::ranges::views;
+            return actions  //
+                   | views::transform([](auto& a) { return detail::to_omit_ptr(a); })
+                   | views::filter([](const auto* ptr) { return ptr != nullptr; })
+                   | views::transform([](auto* ptr) -> decltype(auto) { return *ptr; });
+        }
+
+    }
+
+    inline auto Solution::packages_to_omit() const
+    {
+        return detail::packages_to_omit_impl(actions);
     }
 
-    // TODO(C++20): Poor man's replacement to range filter transform
-    template <typename Iter, typename UnaryFunc>
-    void for_each_to_omit(Iter first, Iter last, UnaryFunc&& func)
+    inline auto Solution::packages_to_omit()
     {
-        for (; first != last; ++first)
+        return detail::packages_to_omit_impl(actions);
+    }
+
+    namespace detail
+    {
+        template <typename Action>
+        constexpr auto package_unique_ptrs(Action& action)
         {
-            if (auto* const ptr = detail::to_omit_ptr(*first))
+            auto out = std::array{
+                to_omit_ptr(action),
+                to_install_ptr(action),
+                to_remove_ptr(action),
+            };
+            for (std::size_t i = 1; i < out.size(); ++i)
             {
-                if constexpr (std::is_same_v<decltype(func(*ptr)), util::LoopControl>)
+                for (std::size_t j = i + 1; j < out.size(); ++j)
                 {
-                    if (func(*ptr) == util::LoopControl::Break)
+                    if (out[j] == out[i])
                     {
-                        break;
+                        out[j] = nullptr;
                     }
                 }
-                else
-                {
-                    func(*ptr);
-                }
             }
+            return out;
         }
+
+        template <typename Range>
+        auto packages_impl(Range& actions)
+        {
+            namespace views = std::ranges::views;
+            return actions                                                             //
+                   | views::transform([](auto& a) { return package_unique_ptrs(a); })  //
```

Comment:
Potential very minor improvement, can be ignored: the lambda wouldnt be necessary if such algorithm (`package_unique_ptrs`) were implemented as a callble object (with a template call operator member) instead of a template free function.